plutus-core/testlib/PlutusCore/Generators/PIR.hs

```diff
@@ -0,0 +1,1752 @@
+ifSizeZero :: GenTm a -> GenTm a -> GenTm a
+ifSizeZero ifZ nonZ = do
+  n <- asks geSize
+  if n <= 0 then ifZ else nonZ
+
+-- | Locally set the size in a generator
+withSize :: Int -> GenTm a -> GenTm a
+withSize = onSize . const
+
+-- | Split the size between two generators in the ratio specified by
+-- the first two arguments.
+sizeSplit_ :: Int -> Int -> GenTm a -> GenTm b -> (a -> b -> c) -> GenTm c
+sizeSplit_ a b ga gb = sizeSplit a b ga (const gb)
+
+-- | Split the size between two generators in the ratio specified by
+-- the first two arguments and use the result of the first generator
+-- in the second.
+sizeSplit :: Int -> Int -> GenTm a -> (a -> GenTm b) -> (a -> b -> c) -> GenTm c
+sizeSplit a b ga gb f = do
+  n <- asks geSize
+  let na = (a * n) `div` (a + b)
+      nb = (b * n) `div` (a + b)
+  x <- withSize na ga
+  f x <$> withSize nb (gb x)
+
+-- * Dealing with fresh names
+
+-- | Get the free variables of a term
+fvTerm :: Term TyName Name DefaultUni DefaultFun ()
+       -> Set Name
+fvTerm tm = case tm of
+  Let _ Rec binds body -> Set.unions
+    (fvTerm body : [ fvTerm body | TermBind _ _ _ body <- toList binds ])
+    `Set.difference` Map.keysSet (foldr addTmBind mempty binds)
+  Let _ _ binds body -> foldr go (fvTerm body) binds
+    where go (TermBind _ _ (VarDecl _ x _) body) free = fvTerm body <> Set.delete x free
+          go _ free                                   = free
+  Var _ nm       -> Set.singleton nm
+  TyAbs _ _ _ t  -> fvTerm t
+  LamAbs _ x _ t -> Set.delete x (fvTerm t)
+  Apply _ t t'   -> fvTerm t <> fvTerm t'
+  TyInst _ t _   -> fvTerm t
+  Constant{}     -> mempty
+  Builtin{}      -> mempty
+  Error{}        -> mempty
+  IWrap{}        -> error "fvTerm: IWrap"
+  Unwrap{}       -> error "fvTerm: Unwrap"
+
+-- | Get the free variables in a type that appear in negative position
+negativeVars :: Type TyName DefaultUni () -> Set TyName
+negativeVars ty = case ty of
+  TyFun _ a b      -> positiveVars a <> negativeVars b
+  TyApp _ a b      -> negativeVars a <> negativeVars b
+  TyLam _ x _ b    -> Set.delete x $ negativeVars b
+  TyForall _ x _ b -> Set.delete x $ negativeVars b
+  TyVar _ _        -> mempty
+  TyBuiltin{}      -> mempty
+  TyIFix{}         -> error "negativeVars: TyIFix"
+
+-- | Get the free variables in a type that appear in positive position
+positiveVars :: Type TyName DefaultUni () -> Set TyName
+positiveVars ty = case ty of
+  TyFun _ a b      -> negativeVars a <> positiveVars b
+  TyApp _ a b      -> positiveVars a <> positiveVars b
+  TyLam _ x _ b    -> Set.delete x $ positiveVars b
+  TyForall _ x _ b -> Set.delete x $ positiveVars b
+  TyVar _ x        -> Set.singleton x
+  TyBuiltin{}      -> mempty
+  TyIFix{}         -> error "positiveVars: TyIFix"
+
+-- | Get the free type variables in a type along with how many
+-- times they occur. The elements of the map are guaranteed to be
+-- non-zero.
+fvTypeBag :: Type TyName DefaultUni () -> Map TyName Int
```

Comment:
You and I have different definitions of the word "easily". The Merriam-Webster definition reads "easily: in an easy manner : without difficulty" - which I think is closer to my definition in this instance ;)

Reply:
Easily if you're a little familiar with lens :) and in this codebase we use lenses a bit, and specifically we use them for tree traversals a lot. So if we're going to add more such things, we should add them in the same way as the existing things, otherwise the codebase as a whole gets harder!

Reply:
@michaelpj 

> Hmm, I think probably what we want is to make fvType and friends into Folds (from lens) and then you can define this fairly easily in terms of that.

I recall you've implemented support for that. Is that right? Can it be used to make `fvTypeBag` a one-liner?

@MaximilianAlgehed 

> I'm not sure writing this as a generic traversal helps - you're still stuck having to specify which cases end up being "delete this thing" and which cases are "union with +". How does generics help make this code more readable?

As far as I understand, you only need to do it once in the definition of the fold. After that you can instantiate it to get `fvType` or the "Bag" version of it in a fairly trivial manner.

Reply:
Nearly a one-liner, something like this:
```haskell
-- copied from the definition of 'setOf', it's identical just with a
-- different 'singleton'
-- | Create a 'MultiSet' from a 'Getter', 'Fold', etc.
multiSetOf :: Getting (MS.MultiSet a) s a -> s -> MS.MultiSet a
multiSetOf l = views l MS.singleton

freeVarsSet' :: Term -> S.Set Name
freeVarsSet' = setOf freeVars
freeVarsMultiSet :: Term -> MS.MultiSet Name
freeVarsMultiSet = multiSetOf freeVars
```

Reply:
Okay, `fvTypeBag = toMap . multiSetOf ftvTy` seems to work. Thanks!

Reply:
You could also just use `MultiSet` throughout instead of `Map a Integer`, but :shrug: